src/internal/stream/heap/heap.go

```diff
@@ -0,0 +1,90 @@
+package heap
+
+type Heap[T any] struct {
+	x  []T
+	lt func(a, b T) bool
+}
+
+func New[T any](lt func(a, b T) bool) Heap[T] {
+	return Heap[T]{}
+}
+
+func (h *Heap[T]) Push(x T) {
+	h.x = Push(h.x, x, h.lt)
+}
+
+func (h *Heap[T]) Pop() (ret T, exists bool) {
+	if len(h.x) == 0 {
+		return ret, false
+	}
+	ret, h.x = Pop(h.x, h.lt)
+	return ret, true
+}
+
+func (h *Heap[T]) Peek() (ret T, exists bool) {
+	if len(h.x) == 0 {
+		return ret, false
+	}
+	return Peek(h.x), true
+}
+
+func (h *Heap[T]) Len() int {
+	return len(h.x)
+}
+
+// Push adds x to the heap h
+func Push[E any, S ~[]E](h S, x E, lt func(a, b E) bool) S {
```

Comment:
Why do push and pop take lt, but the h also contains an lt added at New time?

Reply:
Push and Pop are for using a slice as a heap directly.  The Heap type is an opaque container type.